Argo/Globals/Parser.swift

```diff
@@ -0,0 +1,26 @@
+import Box
+
+public enum Parser<T> {
+  case Success(Box<T>)
+  case TypeMismatch(String)
+  case MissingKey(String)
+
+  public var value: T? {
+    switch self {
+    case let .Success(box): return box.value
+    default: return .None
+    }
+  }
+
+  public static func typeMismatch(expectedType: String, object: JSON) -> Parser<T> {
+    return .TypeMismatch("\(object) is not a \(expectedType)")
+  }
+}
+
+func parseOptional<A>(p: Parser<A>) -> Parser<A?> {
```

Comment:
Right. Picture that the response contains an optional email address, and then the server switches to sending an array of email addresses.

If you fail on type mismatch, you'll catch that immediately.

If you ignore type mismatch, your parser will claim that it's cool every time and you'll wonder why the server isn't sending any usernames.